oauth2client/flask.py

```diff
@@ -0,0 +1,299 @@
+                'OAUTH2_CLIENT_ID' not in app.config):
+            raise ConfigurationError(
+                "OAuth2 configuration could not be found. "
+                "Either OAUTH2_CLIENT_SECRETS_JSON or OAUTH2_CLIENT_ID and "
+                "OAUTH2_CLIENT_SECRET must be specified.")
+
+        return (app.config['OAUTH2_CLIENT_ID'],
+                app.config['OAUTH2_CLIENT_SECRET'])
+
+    def _update_session(self, resp):
+        """Ensures that the session's copy of the credentials is updated
+        when the credential refreshes its access token."""
+        ctx = _app_ctx_stack.top
+        if hasattr(ctx, 'oauth2_credentials'):
+            session['oauth2credentials'] = ctx.oauth2_credentials.to_json()
+        return resp
+
+    def _request_user_info(self):
+        """
+        Makes an HTTP request to the Google+ API to retrieve the user's basic
+        profile information, including full name and photo.
+        """
+        http = self.http()
+        resp, content = http.request(
+            'https://www.googleapis.com/plus/v1/people/me')
+
+        if resp.status != 200:
+            current_app.logger.error(
+                "Error while obtaining user profile: %s" % resp)
+            return None
+
+        profile = json.loads(content)
+        return profile
+
+    def _make_flow(self):
+        # Generate a CSRF token to prevent malicious requests.
+        state = ''.join(random.choice(string.ascii_uppercase + string.digits)
+                        for x in range(32))
+
+        session['oauth2state'] = state
+
+        return OAuth2WebServerFlow(
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+            scope=self.scopes,  # TODO
+            state=state,
+            approval_prompt='force',  # TODO
+            redirect_uri=url_for("oauth2.callback", _external=True))
+
+    def _create_blueprint(self):
+        bp = Blueprint('oauth2', __name__)
+        bp.add_url_rule('/oauth2authorize', 'authorize', self.authorize_view)
+        bp.add_url_rule('/oauth2logout', 'logout', self.logout_view)
+        bp.add_url_rule('/oauth2callback', 'callback', self.callback_view)
+
+        return bp
+
+    def authorize_view(self):
+        return_url = request.args.get('return_url')
+        if not return_url:
+            return_url = request.referrer or '/'
+        session['oauth2return'] = return_url
+
+        flow = self._make_flow()
+        auth_url = flow.step1_get_authorize_url()
+        return redirect(auth_url)
+
+    def logout_view(self):
+        session.pop('oauth2credentials', None)
+        session.pop('oauth2profile', None)
+        session.pop('oauth2userid', None)
+
+        if self.logout_callback:
+            self.logout_callback()
+
+        return redirect(request.referrer or '/')
+
+    def callback_view(self):
+        # Check the CSRF token.
+        if request.args.get('state', '') != session.pop('oauth2state', None):
```

Comment:
'' is not equal to False, thus preventing the situation where both states are undefined and passes the check.

Alternatively, it's probably better to explicitly check for the existence of these keys and return a 400.